### PR TITLE
feat: incentive APY query and rewards height query

### PR DESCRIPTION
**Pull request by toteki**
Adds oracle-enabled APY query to incentive module, as well as a query for last rewards height which will help debug testnet (and we can keep afterwards)

Also fixes the program-starting bug found on testnet, and adds some tests

### CODE REVIEW

File: x/incentive/keeper/grpc_query.go [hunk 1]
```diff
@@ -277,3 +278,77 @@ func (q Querier) CurrentRates(
...
+	// that uToken denom, by the ratio of the total bonded amount to the reference amount.
+	bonded := k.getTotalBonded(ctx, req.UToken)
+	rewards := sdk.NewCoins()
+	exponent := k.getRewardAccumulator(ctx, req.UToken).Exponent
+	for _, p := range programs {
+		if p.UToken == req.UToken {
+			// seconds per year / duration = programsPerYear (as this query assumes incentives will stay constant)
+			programsPerYear := sdk.MustNewDecFromStr("31557600").Quo(sdk.NewDec(p.Duration))
+			// reference amount / total bonded = rewardPortion (as the more uTokens bond, the fewer rewards each earns)
+			rewardPortion := ten.Power(uint64(exponent)).QuoInt(bonded.Amount)
```
Comment (robert-zaremba) [inline x/incentive/keeper/grpc_query.go]: let's move it out of the loop. Or ideally, we have a list of all sane power of tens in this module, or in utils.
```
var Powers10 []sdk.Dec
```

File: x/incentive/keeper/grpc_query.go [hunk 2]
```diff
@@ -277,3 +278,77 @@ func (q Querier) CurrentRates(
...
+				p.TotalRewards.Denom,
+				programsPerYear.Mul(rewardPortion).MulInt(p.TotalRewards.Amount).TruncateInt(),
+			)
+			// add this program's annual rewards to the total for all programs incentivizing this uToken denom
+			rewards = rewards.Add(rewardCoin)
+		}
+	}
+
+	// compute oracle price ratio of rewards to reference bond amount
+	referenceUToken := sdk.NewCoin(req.UToken, ten.Power(uint64(exponent)).TruncateInt())
```
Comment (robert-zaremba) [inline x/incentive/keeper/grpc_query.go]: do we need "reference" prefix in this variable (and below)?

File: x/incentive/keeper/grpc_query.go [hunk 3]
```diff
@@ -277,3 +278,77 @@ func (q Querier) CurrentRates(
...
+	// to compute the rewards a reference amount (10^exponent) of bonded uToken is currently earning,
+	// we need to divide the total rewards being distributed by all ongoing incentive programs targeting
+	// that uToken denom, by the ratio of the total bonded amount to the reference amount.
+	bonded := k.getTotalBonded(ctx, req.UToken)
+	rewards := sdk.NewCoins()
+	exponent := k.getRewardAccumulator(ctx, req.UToken).Exponent
+	for _, p := range programs {
+		if p.UToken == req.UToken {
+			// seconds per year / duration = programsPerYear (as this query assumes incentives will stay constant)
+			programsPerYear := sdk.MustNewDecFromStr("31557600").Quo(sdk.NewDec(p.Duration))
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/incentive/keeper/grpc_query.go]: * let's use `sdk.NewDec(..)` 
* `31557600 != 356 * 24*3600 = seconds_per_year`

File: x/incentive/keeper/mock_test.go [hunk 4]
```diff
@@ -191,3 +191,44 @@ func (m *mockLeverageKeeper) GetTokenSettings(_ sdk.Context, denom string) (leve
 	}
 	return leveragetypes.Token{}, leveragetypes.ErrNotRegisteredToken
 }
+
+// TotalTokenValue implements the expected leverage keeper, with UMEE, ATOM, and DAI registered.
+func (m *mockLeverageKeeper) TotalTokenValue(_ sdk.Context, coins sdk.Coins, _ leveragetypes.PriceMode) (sdk.Dec, error) {
+	total := sdk.ZeroDec()
+
+	var (
+		umeePrice = sdk.MustNewDecFromStr("4.21")
```
Comment (robert-zaremba) [inline x/incentive/keeper/mock_test.go:200]:
```suggestion
	var (
		total = sdk.ZeroDec()
		umeePrice = sdk.MustNewDecFromStr("4.21")
```

File: x/incentive/keeper/mock_test.go [hunk 5]
```diff
@@ -191,3 +191,44 @@ func (m *mockLeverageKeeper) GetTokenSettings(_ sdk.Context, denom string) (leve
...
+			)
+
+		case leveragefixtures.AtomDenom:
+			total = total.Add(
+				atomPrice.MulInt(coin.Amount).QuoInt64(1_000000),
+			)
+		case leveragefixtures.DaiDenom:
+			total = total.Add(
+				daiPrice.MulInt(coin.Amount).QuoInt64(1_000000_000000_000000),
+			)
```
Comment (robert-zaremba) [inline x/incentive/keeper/mock_test.go:219]:
```suggestion
			total = total.Add(
				umeePrice.MulInt(coin.Amount).QuoInt64(1_000000))
		case leveragefixtures.AtomDenom:
			total = total.Add(
				atomPrice.MulInt(coin.Amount).QuoInt64(1_000000))
		case leveragefixtures.DaiDenom:
			total = total.Add(
				daiPrice.MulInt(coin.Amount).QuoInt64(1_000000_000000_000000))
```

**Review by robert-zaremba — COMMENTED** [review]
let's also add a unit test to `ActualRates` before merging.

File: x/incentive/keeper/grpc_query.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (toteki) [reply x/incentive/keeper/grpc_query.go]: 31557600 = 86400 * 365.25

File: x/incentive/keeper/grpc_query.go (hunk 1 above)
Comment (toteki) [reply x/incentive/keeper/grpc_query.go]: done

File: x/incentive/keeper/grpc_query.go (hunk 2 above)
Comment (toteki) [reply x/incentive/keeper/grpc_query.go]: keeping for now for clarity (it's not any user's actual bonded amount)

File: proto/umee/incentive/v1/query.proto [hunk 6]
```diff
@@ -217,4 +231,29 @@ message QueryCurrentRatesResponse {
...
+  string uToken = 1;
+}
+
+// QueryActualRatesResponse defines the response structure for the ActualRates gRPC service handler.
+message QueryActualRatesResponse {
+  // APY is the oracle price-adjusted APY of the bonded uToken.
+  string APY = 1 [
+    (gogoproto.nullable)   = false,
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec"
+  ];
```
Comment (robert-zaremba) [inline proto/umee/incentive/v1/query.proto:258]:
```suggestion
    cosmos.base.v1beta1.DecCoin APY = 1 [(gogoproto.nullable)   = false;
```

**Review by robert-zaremba — COMMENTED** [review]
- [ ] could you add a simple test for `ActualRates`? (more tests can follow in a separate task)

File: proto/umee/incentive/v1/query.proto (hunk 6 above)
Comment (toteki) [reply proto/umee/incentive/v1/query.proto:258]: APY here is meant to be a percent (based on usd value), not token

Comment (toteki) [reply proto/umee/incentive/v1/query.proto:258]: for more detail:

say bonding $100 of UMEE results in $10 of ATOM per year at current rates. The return for `ActualRates` is `0.1` (10%) whereas the return of `CurrentRates` specifies an UMEE reference amount and a resulting ATOM amount in tokens

Comment (robert-zaremba) [reply proto/umee/incentive/v1/query.proto:258]: misread it, I thought you cast to DecCoin

**Review by robert-zaremba — APPROVED** [review]

**Pull request merged**